### PR TITLE
feat(ml): pooled multi-asset volatility model with asset embedding (M3)

### PR DESCRIPTION
**Pull request by jsboige**
## Summary
- Implements pooled multi-asset LSTM volatility model (10 crypto coins, 16-dim asset embedding) — Issue #834, Mission 3
- Panel walk-forward validation (WalkForwardSplitter on DATE axis, all assets share same time splits)
- Diebold-Mariano significance test with Newey-West correction
- BEATS/NO BEATS/INCONCLUSIVE verdict logic (≥60% assets better + mean delta >5% + cross-seed consistency ≥0.75)

## Key changes
- **New**: `scripts/train_pooled_volatility.py` (~830 lines) — Full pipeline: data loading, panel features (9 per asset), per-asset normalization, PyTorch LSTM+embedding, panel WF, DM test, GARCH comparison, verdict
- **New**: `tests/ml/test_pooled_volatility.py` (22 tests) — PanelDataset, DM test, features, normalization, model shape, verdict logic

## Architecture
```
Input: [past_features (60d window), asset_id]
  → Asset Embedding (16-dim learnable)
  → LSTM encoder (64 hidden, 2 layers)
  → Concatenate [lstm_output, asset_embedding]
  → MLP head → log(RV) prediction
```

## Test plan
- [x] 22 unit tests pass (PanelDataset, DM test, features, normalization, model, verdict)
- [x] 11 GARCH baseline tests still pass (no regression from M1)
- [x] `--dry-run` completes end-to-end with synthetic data (33/33 tests total)
- [ ] Full run with real data (10 coins, h=5, 4 seeds, 5 folds) — deferred to M4 re-run

## Dependencies
- Requires M1 (PR #838 — GARCH rolling baseline) to be merged first for GARCH comparison
- M4 (re-run Stage 0/1/2 on corrected protocol) depends on this PR + M1

Co-Authored-By: Claude Opus 4.7 <noreply@anthropic.com>

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
## Review po-2025 — PR #839 (M3 Pooled Multi-Asset)

### Vérification des 5 points (CLAUDE.md sec. B)

| # | Point | Verdict |
|---|-------|---------|
| 1 | **Scope réel** | OK — pooled multi-asset LSTM + asset embedding, rien de plus |
| 2 | **Validation auto** | OK — 22 tests passés localement (PanelDataset, DM, features, verdict) |
| 3 | **Cohérence pédagogique** | N/A (code ML, pas notebook) |
| 4 | **Exécution réelle** | OK — `--dry-run` end-to-end vérifié dans le body |
| 5 | **Regression check** | OK — `fit_garch()` dans train_volatility_garch_dl.py passe du leaky au rolling (correction, pas régression) |

### Anti-regression check
- `train_volatility_garch_dl.py`: 6 additions, 22 deletions — remplace l'ancien `fit_garch` leaky (fit in-sample + `cond_var * h`) par un appel à `fit_garch_rolling` (PR #838). **C'est un FIX de data leak, pas une régression.**
- `garch_baseline.py` listé comme ADDED ici mais déjà merged via M1 (#838). Conflit mineur mineur — à vérifier que le merge résout les doublons.

### Critères G (anti-complaisance)
- **G.4 Composite**: 1579 additions, 7 files, 1 feature → OK (sous les seuils)
- **G.3 Test plan**: 3/4 checkés, le 4e (`Full run deferred to M4`) est explicitement reporté → OK
- **G.2 ML metrics**: dry-run only dans ce PR, verdict full run reporté à M4 → cohérent

### Remarque
Le fichier `garch_baseline.py` apparaît comme ADDED dans le diff de cette PR alors qu'il est déjà sur main via M1 (#838). Probablement un artefact de rebase. Le merge devrait résoudre automatiquement.

**Verdict**: Pas de CHANGES_REQUESTED de ma part. PR solide, tests verts, scope maîtrisé.

**Review by clusterManager-Myia — COMMENTED** [review]
## Review ClusterManager — PR #839 (M3 Pooled Multi-Asset Volatility)

### Scope
7 files, 1579 additions. Pooled LSTM with asset embedding + GARCH rolling baseline + DM test. One existing file modified (`train_volatility_garch_dl.py`: 6 additions, 22 deletions — data leak fix).

### garch_baseline.py (252L)
- `fit_garch_leaky`: correctly labeled as legacy/data leak. Kept for comparison only.
- `fit_garch_rolling`: proper rolling refit with `arch_model.forecast()`. Uses `convergence_flag != 0` to skip failed fits — good defensive practice. Rescales %² → original units. Multi-step forecast sums daily variance forecasts (not simple `cond_var * h` scaling) — this is the correct approach for h-step GARCH forecasts.
- `compare_baselines`: clean A/B comparison on last 20% test set.
- One note: `train_window=1000` with `refit_freq=5` means ~200 refits for a 1500-day series. Each refit fits GARCH from scratch — computationally expensive but correct for avoiding look-ahead.

### train_pooled_volatility.py (828L)
- **Architecture**: Asset embedding (16-dim) → concatenate with features → LSTM (64 hidden, 2 layers, dropout 0.1) → MLP head. Clean design.
- **PanelDataset**: per-asset rolling windows, date-range filtering, proper `(features, asset_id, target)` tuple output.
- **Normalization**: per-asset z-score normalization with stats dict for inverse transform. Correct approach for cross-asset pooling.
- **Walk-forward**: date-axis splits (all assets share same time splits). Multi-seed (default 4 seeds).
- **Early stopping**: on training loss with patience=10. Monitoring train loss (not validation) is suboptimal — could overfit. However, the walk-forward structure itself provides OOS evaluation, so this is acceptable as a pragmatic choice.
- **Verdict logic**: BEATS requires ≥60% assets better AND >5% mean delta AND ≥75% cross-seed consistency. Conservative thresholds — good.
- **DM test**: inline implementation (not using M4's `diebold_mariano.py`). Uses biased autocovariance which differs from M4's `np.dot(x[k:], x[:-k]) / n`. Both are valid HAC estimators but the inconsistency between M3 and M4 is worth noting — they may give slightly different DM statistics for the same data.

### train_volatility_garch_dl.py (6+/22-)
- `fit_garch()` now delegates to `garch_baseline.fit_garch_rolling`. Clean fix — old leaky implementation preserved as `fit_garch_leaky`. This is a **data leak fix**, not a regression.

### Tests (172 + 321 = 493L)
- GARCH tests: OOS forecast NaN in train region, no future data (truncation test), multi-step ≠ simple scaling, positive forecasts, short series, compare baselines differ, dry-run.
- Pooled tests: PanelDataset construction, shape, date filtering, multi-asset, DM test (clearly better / equal / worse / multi-horizon), panel features (shape, columns, targets, no NaN), normalization (per-asset zero-mean unit-var, global, stats dict), model forward pass shape, verdict logic (BEATS / NO BEATS / INCONCLUSIVE / empty).
- The no-future-data test (truncation comparison) is particularly valuable.

### Observations
1. **DM implementation duplication**: M3 has its own `diebold_mariano_test` inline (simpler, no Bonferroni, no small-sample correction). M4 (#845) has the full version. After both merge, consider deduplicating — M3 should import from M4's module.
2. **Early stopping on train loss**: Suboptimal but acceptable given walk-forward OOS evaluation.

**COMMENT** — no blocking issues found. Consider DM dedup post-merge.
